apollo-router/src/plugins/telemetry/apollo_otlp_exporter.rs

```diff
@@ -0,0 +1,140 @@
+                ),
+                KeyValue::new("apollo.graph.ref", apollo_graph_ref.to_string()),
+                KeyValue::new("apollo.schema.id", schema_id.to_string()),
+                KeyValue::new(
+                    "apollo.user.agent",
+                    format!(
+                        "{}@{}",
+                        std::env!("CARGO_PKG_NAME"),
+                        std::env!("CARGO_PKG_VERSION")
+                    ),
+                ),
+                KeyValue::new("apollo.client.host", hostname()?),
+                KeyValue::new("apollo.client.uname", get_uname()?),
+            ]),
+            intrumentation_library: InstrumentationLibrary::new(
+                GLOBAL_TRACER_NAME,
+                Some(format!(
+                    "{}@{}",
+                    std::env!("CARGO_PKG_NAME"),
+                    std::env!("CARGO_PKG_VERSION")
+                )),
+                Option::<String>::None,
+                None,
+            ),
+            otlp_exporter: Arc::new(Mutex::new(
+                SpanExporterBuilder::from(
+                    opentelemetry_otlp::new_exporter()
+                        .tonic()
+                        .with_timeout(batch_config.max_export_timeout)
+                        .with_endpoint(endpoint.to_string())
+                        .with_metadata(metadata),
+                        // TBD(tim): figure out why compression seems to be turned off on our collector
+                        // .with_compression(opentelemetry_otlp::Compression::Gzip),
+                )
+                .build_span_exporter()?,
+            )),
+            // TBD(tim): do we need another batch processor for this?
+            // Seems like we've already set up a batcher earlier in the pipe but not quite sure.
+        });
+    }
+
+    pub(crate) fn prepare_for_export(&self, span: &LightSpanData) -> SpanData {
+        SpanData {
+            span_context: SpanContext::new(
+                span.trace_id,
+                span.span_id,
+                TraceFlags::default().with_sampled(true),
+                true,
+                TraceState::default(),
+            ),
+            parent_span_id: span.parent_span_id,
+            span_kind: span.span_kind.clone(),
+            name: span.name.clone(),
+            start_time: span.start_time,
+            end_time: span.end_time,
+            attributes: span.attributes.clone(),
+            events: EvictedQueue::new(0),
+            links: EvictedQueue::new(0),
+            status: Status::Unset,
+            resource: Cow::Owned(self.resource_template.to_owned()),
+            instrumentation_lib: self.intrumentation_library.clone(),
+        }
+    }
+
+    pub(crate) fn export(&self, spans: Vec<SpanData>) -> BoxFuture<'static, ExportResult> {
+        let mut exporter = self.otlp_exporter.lock().unwrap();
+        exporter.export(spans)
+    }
+
+    pub(crate) fn shutdown(&self) {
```

Comment:
If it's a shutdown and we don't use the exporter itself after maybe we should take ownership ? Just a suggestion I don't know if it's valid
```suggestion
    pub(crate) fn shutdown(self) {
```

Reply:
I gave it a try but the compiler was not a fan.